### DISCUSSION TITLE
Generate a documentation page in Markdown from a json description

### DISCUSSION THREAD

**Discussion by mquandalle** [facebook/docusaurus#6009] [Q&A]
Hello,

I have a JSON file describing an API with attributs like `title` and `description`, and would like to render a page in my Docusaurus documentation listing all these API. The description are written is markdown but I was unable to find a `<Markdown />` component that I can use.

Essentialy I would like to display a generated markdown document like so

```jsx
// docs/api.jsx

import apiDescription from "../apiDescription.js"

export default function() {
  return <Markdown>
    {apiDescription.map(({title, description}) => `## ${title} \n\n ${description}`).join('\n\n')}
  </Markdown>
}
```

What is the recommanded way to generate a markdown page from an external data source? Also I'd would like to display a sidebar with all generated headings like for other pages.

**Comment by slorber** [comment]
We don't include any markdown component by default in Docusaurus because such libs add some decent size to the final JS bundle and parsing markdown syntax in the frontend also has a runtime cost that can slow down your page transitions.

If you really want to do this you can add a lib like https://github.com/remarkjs/react-markdown yourself on pages. It adds weight but it might be ok for your site, it's up to you.

Other more lightweight possibilities:
- generate a `apiDescription.md` file at build time
- generate a `apiDescription.json` Markdown AST file at build time and then render it with a markdown renderer (ie, the markdown syntax parser does not need to be sent to the browser)


Note you do not need to use Markdown, you might as well create a comp that manually renders your API:

```jsx
export default function MyAPIComponent() {
  return <>
    {apiDescription.map(({title, description}) => (
      <div>
		    <h2>${title}</h2>
		    <div>{description}</div>
	  </div>
  ))}
  </>
}
```

only problem is that Docusaurus wouldn't be able to compute a table of contents though

**Reply by mquandalle** [reply]
I guess the upcoming [React server components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html) could be a long-term solution?

**Comment by laem** [comment]
@slorber Investigating the build time solution : I can't find any documented way in Docausorus to transform files at build time, e.g. with a simple node script that I would list somewhere, that would be picked at build time with hot reload support. 


Does it involves writing a simple plugin ? https://docusaurus.io/fr/docs/using-plugins

Do I have to override the package.json scripts ? That looks quite complicated.

**Reply by mquandalle** [reply]
For example it looks like Supabase is generating their [api pages](https://supabase.com/docs/reference/javascript/auth-signup) with custom-made build scripts in https://github.com/supabase/supabase/tree/master/web/spec. This indeed looks complicated and ought to be documented.

**Reply by Josh-Cena** [reply]
@laem Yes, for now you would write a plugin yourself to transform the file. However, for now we don't guarantee the execution order of plugins (they are called in parallel) so by the time your plugin has loaded content and created the file, the docs plugin may have finished loading its data as well, causing your generated file to not be picked up. In that case you would either need to render the data entirely with your plugin and not rely on any other plugin, or try to enhance the docs plugin, as described here: https://github.com/facebook/docusaurus/issues/4492#issuecomment-805041391
